src/shell/shell_misc.cpp

```diff
@@ -389,106 +392,39 @@ void DOS_Shell::InputCommand(char * line) {
-						if (*r == ' ') spaces++;
-						else chars++;
-						r++;
-					}
-				}
-				if (*r == '%' && ((spaces > 0 && chars == 0) || (spaces == 0 && chars > 0))) {
-					std::string temp2;
-					/* valid name found. substitute */
-					*r++ = 0; /* ASCIIZ snip */
-					if (GetEnvStr(name,temp2)) {
-						size_t equ_pos = temp2.find_first_of('=');
-						if (equ_pos != std::string::npos) {
-							const char *base = temp2.c_str();
-							const char *value = base + equ_pos + 1;
-							const char *fence = base + temp2.length();
-							assert(value >= base && value <= fence);
-							size_t len = (size_t)(fence-value);
-							if ((w+len) > wf) goto overflow;
-							memcpy(w,value,len);
-							w += len;
-						}
-					}
-				}
-				else {
-					/* nope. didn't find a valid name */
-					while (*r == ' ') r++; /* skip spaces */
-					name--; /* step "name" back to cover the first '%' we found */
-					for (const char* c = name; c < r;) {
-						if (w >= wf) goto overflow;
-						*w++ = *c++;
-					}
-				}
-			}
+/* Note: Buffer pointed to by "line" must be at least CMD_MAXLINE+1 bytes long! */
+void DOS_Shell::ProcessCmdLineEnvVarStitution(char* line) {
+	std::string text = line;
+	const static std::regex re("\\%([^%0-9][^%]*)?%");
+	std::smatch match;
+	/* Iterates over potential %var1%, %var2%, etc matches found in the text string */
+	while (std::regex_search(text, match, re)) {
+        /* "variable_name" and "variable value" store the name and value of the variable(s) */ 
+		std::string variable_name;
+		variable_name = match[1].str();
+		if (!variable_name.size()) {
+			/* Replace %% with the character "surrogate_percent", then (eventually) % */
+			text.replace(match[0].first, match[0].second, std::string(1, surrogate_percent));
+			continue;
+		}
+		while (variable_name.size() && isspace(variable_name.front()))
+			variable_name.erase(0, 1);
+		std::string variable_value;
+		if (variable_name.size() && GetEnvStr(variable_name.c_str(), variable_value)) {
+			const size_t equal_pos = variable_value.find_first_of('=');
+			/* Replace the original %var% with its corresponding value from the environment */
+			if (equal_pos != std::string::npos)
+				text.replace(match[0].first, match[0].second, variable_value.substr(equal_pos+1));
+			else
+				text.replace(match[0].first, match[0].second, "");
 		}
 		else {
-			if (w >= wf) goto overflow;
-			*w++ = *r++;
+			text.replace(match[0].first, match[0].second, "");
 		}
 	}
-	/* complete the C-string */
-	assert(w <= wf);
-	*w = 0;
-	/* copy the string back over the buffer pointed to by line */
-	{
-		size_t out_len = (size_t)(w+1-temp); /* length counting the NUL too */
-		assert(out_len <= CMD_MAXLINE);
-		memcpy(line,temp,out_len);
-	}
-	/* success */
-	return;
-overflow:
-	*line = 0; /* clear string (C-string chop with NUL) */
-	WriteOut("Command input error: string expansion overflow\n");
+	std::replace(text.begin(), text.end(), (char)surrogate_percent, '%');
```

Comment:
Is the `(char)` cast here still needed? (I hope we can drop it)

Reply:
It can be dropped now. Done.